Feature: `--ebs-volumes` option for ephemeral-storage init

jesseanttila-cai opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #389

**Description of changes:**

Adds two new options to `ephemeral-storage init`:

* `--ebs-volumes` specifies EBS volumes to be included in the ephemeral storage array (e.g. `--ebs-volumes /dev/disk/ephemeral-ebs/xvdda /dev/disk/ephemeral-ebs/xvddb`

* `--prefer` specifies combinations of storage types to be included in the ephemeral storage array in order of preference. Currently supported storage types are `ephemeral-disk` and `ebs-volume`. Storage types can be combined with the `+` operator.  For example, `--prefer ephemeral-disk ebs-volume` uses ephemeral disks if available and ebs volumes otherwise. To use a mix of both types, specify `--prefer ephemeral-disk+ebs-volume`.


**Testing done:**

Manual testing of a few different configurations on EKS (see [this comment](https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/395#issuecomment-3224069765))

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

bcressey commented on sources/api/apiserver/src/server/ephemeral_storage.rs:
```diff
@@ -32,35 +32,53 @@ pub struct BindDirs {
-            known_disks
+        };
+        for ebs_volume in ebs_volumes.iter().flatten() {
+            let link_path = format!("/dev/disk/by-ebs-id/{ebs_volume}");
```
Let's add an `ephemeral_ebs_volumes` function, like `ephemeral_devices`, that returns available EBS volumes.

bcressey commented on sources/api/apiserver/src/server/ephemeral_storage.rs:
```diff
@@ -32,35 +32,53 @@ pub struct BindDirs {
+                // if there are no disks specified, and none are available we treat the init as a
+                // no-op to allow "ephemeral-storage init"/"ephemeral-storage bind" to work on instances
+                // with and without ephemeral storage
+                if known_disks.is_empty() && ebs_volumes.is_none() {
+                    info!("no ephemeral disks found, skipping ephemeral storage initialization");
+                    return Ok(());
+                }
```
There are a couple of design questions here that are worth considering, in the context of the patch that I shared that designates the `xvdda` - `xvddx` range for "ephemeral" EBS volumes.

Should we allow both EBS and instance store volumes in the same RAID-0 array? I would default to "no", and reject the API request if we're given both. We can change our minds later, but it seems weird to mix devices like this in the same array.

If both lists are empty, and both instance store disks and ephemeral EBS volumes are present, and we can't mix the two in the same array - then which devices do we include?

Some options:
1. Add another API flag, like `--prefer-ebs-volumes`, which controls the behavior when presented with two empty lists.
    1. If it's true:
        - If ephemeral EBS volumes are present: we'll use them in the array.
        - If no ephemeral EBS volumes are present: we'll use ephemeral disks, if any.
    2. If it's false:
         - If ephemeral disks are present: we'll use them in the array.
         - If no ephemeral disks are present: we'll use ephemeral EBS volumes, if any.
2. Punt on the problem and don't support implicit array creation for EBS volumes.

`--prefer-ebs-volumes` could also solve the first issue - if we're given two lists and told which one to prefer, we can ignore the other one rather than return an error.

(For the actual implemention, I'd favor using an enum and having `--prefer ephemeral-disks` and `--prefer ebs-volumes` rather than a boolean. That will be easier to extend if we end up with another storage category. You can put it in `models/src/ephemeral_storage.rs` so that the client and server can share the same type definition.)

bcressey replied:
With the enum approach, we could also just have all four options as variants:
```
--prefer [ disks-only
         | ebs-volumes-only
         | disks-to-ebs-volumes
         | ebs-volumes-to-disks ]
```

Then the default can be `disks-only` for backwards compatibility.

bcressey commented on sources/api/apiserver/src/server/ephemeral_storage.rs:
```diff
@@ -92,7 +92,13 @@ pub fn initialize(
     let device_name = match disks.len() {
-        1 => disks.first().expect("non-empty").clone(),
+        1 => {
+            let disk_name = disks.first().expect("non-empty").clone();
+            let link_name = format!("{RAID_DEVICE_DIR}{RAID_DEVICE_NAME}");
+            std::os::unix::fs::symlink(disk_name, &link_name)
+                .context(error::DiskSymlinkFailureSnafu)?;
+            link_name
```
Here you're fixing a pre-existing bug - we effectively lost track of which disk we used during the "init" operation. That didn't matter in the case where we use all ephemeral disks, but could have mattered on an instance that had three (`nvme1`, `nvme2`, `nvme3`) where the "init" request specified only one (`nvme2`) so we didn't end up creating an array.

The symlink approach is a nice, cheap way to preserve the state we need, but it can cause trouble to create links in "owned" directories under `/dev`, like `/dev/md` (which `mdadm` "owns").

Instead I'd suggest using [pathdiff::diff_paths](https://docs.rs/pathdiff/latest/pathdiff/fn.diff_paths.html) and creating a relative symlink from `/dev/disk/ephemeral-storage` to either the single disk or EBS volume, or to the RAID-0 array.

bcressey reviewed (commented):
Here's a [patch](https://gist.github.com/bcressey/3db84ff444d073a11f3c5356bf4d8cb4) that populates `/dev/disk/ephemeral-ebs` with symlinks for non-system EBS volumes in the `xvdda` - `xvddx` range.

In combination with my other suggestions, this would let you provide user-data like this to automatically configure up to 24 EBS volumes in a RAID-0 array, if present, and fall back to instance store otherwise:
```
[settings.bootstrap-commands.k8s-ephemeral-storage]
commands = [
    ["apiclient", "ephemeral-storage", "init", "--prefer", "ebs-volumes"],
    ["apiclient", "ephemeral-storage" ,"bind", "--dirs", "/var/lib/containerd", "/var/lib/kubelet", "/var/log/pods"]
]
essential = true
mode = "always"
```

jesseanttila-cai commented:
Thank you kindly for the comments and suggestions. I had some doubts about the handling of system ebs volumes and some edge-cases involving mixed instance storage/ebs volume arrays, which are solved by the `/dev/disk/ephemeral-ebs` links and the `--prefer` flag suggestion.
Unfortunately I won't be able to (read: shouldn't) implement and test any of these changes until I return from holiday on the 25th, but now that I have the AMI build set-up it shouldn't take four months this time.

jesseanttila-cai commented on sources/api/apiclient/src/main.rs:
```diff
@@ -271,7 +272,12 @@ fn usage() -> ! {
             --ebs-volumes VOLUME [VOLUME ..]
-                                       Additional EBS volumes to configure for storage.
+                                       EBS volumes in the `xvdda`-`xvddx` range to configure for storage.
+            --prefer PREFERENCE [PREFERENCE ..]
+                                       Epheremeral storage type preference, in descending order. Allowed
+                                       values: `ephemeral-disk`, `ebs-volume` or a combination of these
+                                       joined by `+`. Defaults to `ephemeral-disk` only. This option is
+                                       ignored if `--disks` or `--ebs-volumes` is set.
```
Instead of the enum solution I went with an ordered list of "preferences", which are sets of allowed storage types. This should be easier to expand in case more storage types are added in the future.

There are effectively five distinct preference options available:
* `--prefer ephemeral-disk`: use only ephemeral instance storage (default)
* `--prefer ebs-volume`: use only ebs volumes
* `--prefer ephemeral-disk ebs-volume`: use ephemeral instance storage if available, otherwise ebs volumes
* `--prefer ebs-volume ephemeral-disk`: same as above except in the opposite order
* `--prefer ephemeral-disk+ebs-volume`: use all available storage of either type

The `--disks` and `--ebs-volumes` options override `--prefer`, which allows for a specific custom storage setup to be defined. This is also backwards compatible with the previous behaviour of the `--disks` option.

bcressey replied:
Makes sense; I agree the enum variants would get unwieldy if more permutations of storage were added, while this will be easy to extend.

jesseanttila-cai commented on sources/models/src/ephemeral_storage.rs:
```diff
@@ -18,12 +18,58 @@ impl Display for Filesystem {
+            ephemeral_disk: false,
+            ebs_volume: false,
+        };
+        for i in value.split("+") {
+            match i {
+                "ephemeral-disk" => preference.ephemeral_disk = true,
+                "ebs-volume" => preference.ebs_volume = true,
+                x => return Err(x),
+            }
```
Should `""` be allowed here? It would allow defining an empty set (not very useful) and superfluous `+`s, e.g. `ephemeral-disk+`. The main benefit would probably be enabling round-tripping through `Display` and `TryFrom<&str>` for the empty set case.

bcressey replied:
The round-trip property is nice, though I don't feel strongly. Up to you!

jesseanttila-cai commented:
Results from manual testing with the latest commit:

| instance config | A | B | C | D | E |
|------------------|---|---|---|---|---|
|m5.large|20G|20G|20G|20G|X|
|m5.large+4x8GB|20G|32G|32G|32G|16G|
|m5.large+16x4GB|20G|64G|64G|64G|8G|
|m5d.large|70G|70G|70G|70G|X|
|m5d.large+4x8GB|70G|32G|70G|102G|16G|
|m5d.large+16x4GB|70G|64G|70G|134G|8G|

where
* **A**: default (equivalent to `--prefer ephemeral-disk`)
* **B**: `--prefer ebs-volume ephemeral-disk`
* **C**: `--prefer ephemeral-disk ebs-volume`
* **D**: `--prefer ephemeral-disk+ebs-volume`
* **E**: `--ebs-volumes /dev/disk/ephemeral-ebs/xvddb /dev/disk/ephemeral-ebs/xvddd`

All test cases reported reasonable values for ephemeral storage capacity, except for two cases where the bootstrap command failed due to missing ebs volumes. A capacity of 20G corresponds to the default EBS volume added by Karpenter, which is used when `ephemeral-storage init` does not find any storage devices of an allowed type.

bcressey reviewed (approved):
Apologies for the delayed review! LGTM.

bcressey commented:
Let me know if you're ready to mark this as "ready for review" and I can kick off CI tests. The only change I'd suggest is squashing the commits together and splitting them back out into a pair of apiclient and apiserver commits.

Commit:
apiserver: add --ebs-volumes and --prefer options to ephemeral-storage init command

Commit:
apiclient: add --ebs-volumes and --prefer options to ephemeral-storage init command

jesseanttila-cai commented:
I've refactored the PR to two commits, the apiserver commit also includes the changes to `models` & `os.spec`. There was also a very minor conflict with commit 3fd2ad12388e3c38bd3997883e6ec5d14ce6234a which has now been resolved.

ytsssun commented on sources/api/apiclient/src/main.rs:
```diff
@@ -269,7 +271,13 @@ fn usage() -> ! {
+            --ebs-volumes VOLUME [VOLUME ..]
+                                       EBS volumes in the `xvdda`-`xvddx` range to configure for storage.
+            --prefer PREFERENCE [PREFERENCE ..]
+                                       Epheremeral storage type preference, in descending order. Allowed
```
Nit: typo
```suggestion
                                       Ephemeral storage type preference, in descending order. Allowed
```

ginglis13 commented on sources/api/apiserver/src/server/ephemeral_storage.rs:
```diff
@@ -298,6 +338,29 @@ pub fn ephemeral_devices() -> Result<Vec<String>> {
 
+/// ephemeral_ebs_volumes returns the full path name to the ebs volumes in /dev/disk/ephemeral-ebs
+pub fn ephemeral_ebs_volumes() -> Result<Vec<String>> {
+    const EPHEMERAL_EBS_PATH: &str = "/dev/disk/ephemeral-ebs";
```
nit: I'd prefer this defined at the top of this file as `static` below `EPHEMERAL_STORAGE_LINK`

ginglis13 commented on sources/api/apiclient/src/main.rs:
```diff
@@ -269,7 +271,13 @@ fn usage() -> ! {
             --disks DISK [DISK ...]    Local disks to configure for storage. Default is all ephemeral
                                        disks.
-
+            --ebs-volumes VOLUME [VOLUME ..]
```
Won't block on it but would be nice to update the README for `apiclient` with these changes

Commit:
apiserver: add list-ebs-volumes subcommand, fix list_ephemeral_storage_dirs doc comment

Commit:
apiclient: add list-ebs-volumes subcommand, improve ephemeral storage documentation

ginglis13 commented:
I performed some more extensive manual testing today, everything LGTM with 2 non-blocking observations:

* missing `list-volumes` sub-command. We have an equivalent `list-disks`, this would be nice to have for parity
* once a preference is set via `--prefer`, it cannot be updated

bcressey commented on sources/api/apiserver/src/server/ephemeral_storage.rs:
```diff
@@ -95,22 +139,21 @@ pub fn initialize(fs: Option<Filesystem>, disks: Option<Vec<String>>) -> Result<
         info!("{device_name:?} is already formatted as {fs}, skipping format");
     }
 
+    // Create link to formatted device for use in `bind`
+    std::os::unix::fs::symlink(&device_name, EPHEMERAL_STORAGE_LINK)
+        .context(error::DiskSymlinkFailureSnafu {})?;
```
`init` now fails after a second invocation:
```
> apiclient ephemeral-storage init --disks /dev/disk/ephemeral/virtio-eph01234
Failed to initialize ephemeral storage:
  Failed POST request to '/actions/ephemeral-storage/init':
    Status 500 when POSTing /actions/ephemeral-storage/init:
      Unable to initialize ephemeral storage:
        Failed to create disk symlink File exists (os error 17)
```

Not sure how reliable a second `init` was in the past but the intent is for it to be idempotent.

Commit:
apiserver: overwrite symlink in ephemeral storage init

jesseanttila-cai commented:
I've added a `list-ebs-volumes` subcommand to match `list-disks`, and corrected the documentation of `list-dirs` to match its actual behavior. I haven't had a chance to test these changes yet, but hopefully will by the end of the week.

bcressey commented:
Testing LGTM:
```
# Attached two EBS volumes which can be listed.
bash-5.1# apiclient ephemeral-storage list-ebs-volumes
/dev/disk/ephemeral-ebs/xvddb
/dev/disk/ephemeral-ebs/xvdda

# Volumes were added to the array.
bash-5.1# apiclient ephemeral-storage init --prefer ebs-volume

bash-5.1# cat /proc/mdstat
Personalities : [raid0]
md127 : active raid0 nvme3n1[1] nvme4n1[0]
      209582080 blocks super 1.2 256k chunks

bash-5.1# nvme amzn id-ctrl /dev/nvme4n1|grep dev
bdev      : /dev/xvddb

bash-5.1# nvme amzn id-ctrl /dev/nvme3n1|grep dev
bdev      : /dev/xvdda

# init operation is now idempotent
bash-5.1# apiclient ephemeral-storage init
```

bcressey commented:
Nice! Thanks for contributing this!
